src/ray/util/ordered_set.h

```diff
@@ -0,0 +1,48 @@
+#ifndef RAY_UTIL_ORDERED_SET_H
+#define RAY_UTIL_ORDERED_SET_H
+
+#include <list>
+#include <unordered_map>
+
+template <typename T>
+class ordered_set {
+ public:
+  ordered_set() {}
+
+  ordered_set(const ordered_set &other) = delete;
+
+  ordered_set &operator=(const ordered_set &other) = delete;
+
+  void push_back(const T &value) {
+    RAY_CHECK(iterators_.find(value) == iterators_.end());
+    auto list_iterator = elements_.insert(elements_.end(), value);
+    iterators_[value] = list_iterator;
+  }
+
+  size_t count(const T &k) const { return iterators_.count(k); }
+
+  void pop_front() {
+    iterators_.erase(elements_.front());
+    elements_.pop_front();
+  }
+
+  const T &front() const { return elements_.front(); }
+
+  size_t size() const noexcept {
+    RAY_CHECK(elements_.size() == iterators_.size());
```

Comment:
You sure elements_.size() is constant time? https://stackoverflow.com/questions/228908/is-listsize-really-on
Seems like it depends on the implementation version.

Reply:
I think only GCC 4.8.1 would support C++11 (to compile the codebase) while simultaneously degenerating to O(N) time? Earlier versions wouldn't fully support C++11, whereas GCC 5 and later should be constant time.

Reply:
We are using the old gcc ABI because of TensorFlow, and I think it is O(n) there, so we should remove this check (or replace it with DCHECK).

Reply:
Oh wow. In that case you might want to either remove `size()` altogether (`empty()` might be sufficient?) or, if not, change the implementation altogether (e.g. perhaps to what I suggested elsewhere) since callers will likely expect this to be constant-time.

Reply:
doing .size on the unordered_map is O(1), so just removing the CHECK is a good option I think :)

Reply:
Fixed! Good catch @ericl!